towhee/dataframe/dataframe.py

```diff
@@ -35,19 +37,171 @@ def __init__(self, name: str, data: List[Tuple[Variable]] = None):
+        return self._total - start
+
+    def set_end(self):
+        with self._lock:
+            self._end = True
+
+    # TODO (junjie.jiangjjj)
+    def _gc(self) -> None:
+        """
+        Delete the data which all registered iter has read
+        """
+        return None
+
+    def map_iter(self):
+        it = MapIterator(self)
+        with self._registered_lock:
+            self._registered_iter.append(it)
+        return it
+
+
+class _BaseIterator:
+    """Base iterator implementation. All iterators should be subclasses of
+    `_BaseIterator`.
+
+    Args:
+        df: The dataframe to iterate over.
+    """
+
+    def __init__(self, df: DataFrame):
+        self._df_ref = weakref.ref(df)
+        self._cur_index = 0
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        # The base iterator is purposely defined to have exatly 0 elements.
+        raise StopIteration
+
+    def accessible_size(self) -> int:
+        """
+        Return current accessible data size
+        """
+        return self._df_ref().get_size(self._cur_index)
+
+
+class MapIterator(_BaseIterator):
+    """Iterator implementation that traverses the dataframe line-by-line.
+
+    Args:
+        df: The dataframe to iterate over.
+    """
+
+    def __next__(self):
+        end, data = self._df_ref().get(self._cur_index, 1)
+        if end and len(data) == 0:
+            raise StopIteration
+        self._cur_index += len(data)
+        return data
+
+
+class BatchIterator:
+    """Iterator implementation that traverses the dataframe multiple rows at a time.
+
+    Args:
+        df:
+            The dataframe to iterate over.
+        size:
+            batch size. If size is None, then the whole variable set will be batched
+            together.
+    """
+
+    def __init__(self, df: DataFrame, size: int = None):
+        super().__init__(df)
+        self._size = size
+
+    def __next__(self):
+        raise NotImplementedError
+
+
+class GroupIterator:
+    """Iterator implementation that traverses the dataframe based on a custom group-by
+    function.
+
+    Args:
+        df:
+            The dataframe to iterate over.
+        func:
+            The function used to return grouped data within the dataframe.
+    """
+
+    def __init__(self, df: DataFrame, func: Callable[[], None]):
+        super().__init__(df)
+        self._func = func
+
+    def __next__(self):
+        raise NotImplementedError
+
+
+class RepeatIterator:
+    """Iterator implementation that repeatedly accesses the first line of the dataframe.
+
+        Args:
+            df:
+                The dataframe to iterate over.
+            n:
+                the repeat times. If `None`, the iterator will loop continuously.
+    """
 
-    def __delete__(self, val):
-        del self.data[val]
+    def __init__(self, df: DataFrame, n: int = None):
+        # self._n = n
+        # self._i = 0
+        raise NotImplementedError
```

Comment:
This iterator looks mostly fine to me, we can remove this error and deliver it with v0.1 (adding tests later).

Reply:
Only implement map iterator, others will implement after v0.1